packages/flutter_tools/lib/src/build_system/build_system.dart

```diff
@@ -775,7 +776,14 @@ class FlutterBuildSystem extends BuildSystem {
     for (final String lastOutput in lastOutputs) {
       if (!currentOutputs.containsKey(lastOutput)) {
         final File lastOutputFile = fileSystem.file(lastOutput);
-        ErrorHandlingFileSystem.deleteIfExists(lastOutputFile);
+        try {
+          ErrorHandlingFileSystem.deleteIfExists(lastOutputFile);
```

Comment:
Isn't the bug here that `ErrorHandlingFileSystem.deleteIfExists()` didn't catch this? From reading the code, I'm guessing the exception was thrown from `file.existsSync()`? I think we should update that function to handle FileSystemExceptions.

Reply:
(I already discussed this with christopherfujino, but I'm responding here for posterity's sake).

> I'm guessing the exception was thrown from `file.existsSync()`?

No, looking at the stack trace reported from the issue, this is coming from a `deleteSync` call. However, you are right that this permissions issue should have been caught by [our existing logic](https://github.com/flutter/flutter/blob/4a65a76279cda479d0c8a2a57ac8f34890e1adf2/packages/flutter_tools/lib/src/base/error_handling_io.dart#L785) in `ErrorHandlingFile`. This means that the `File` instance here was not an instance of `ErrorHandlingFile` (and any instance of `File` within tool code should be a `ErrorHandlingFile`). This is the root of the problem, but troubleshooting this would be non-trivial.

Reply:
Maybe one way to troubleshoot this would be to try to recreate the tool invocation from the stacktrace locally but with an edited tool that checks the runtimetype of File and throws if it's not an ErrorHandlingFile. And then trace from there how you got that file.

Reply:
I did this and ran `flutter build appbundle`, but everything that went through `deleteIfExists` was an `ErrorHandlingFile` or `ErrorHandlingDirectory`. 🤷